File: go/terminal/proto/session.go
package proto

import (
	"encoding/json"
	"fmt"
	"log"
	"math"
	"net/url"
	"sync"
	"time"

	"github.com/nullentropy/caution/go/terminal/gfx"
	"github.com/nullentropy/caution/go/terminal/ui"

	"github.com/gorilla/websocket"
)

type serverMsg struct {
	T         string             `json:"t"`
	Seq       int                `json:"seq"`
	Root      *nodeJSON          `json:"root"`
	SID       string             `json:"sid"`
	Theme     map[string]string  `json:"theme"`
	Metrics   map[string]float64 `json:"metrics"`
	Menu      []MenuSpec         `json:"menu"`
	Title     string             `json:"title"`
	Keys      []keySpec          `json:"keys"`
	Resources *resourceSpec      `json:"resources"`
	Ops       []patchOp          `json:"ops"`
}

// resourceSpec is the mount's preload block (the `resource` op inline);
// images today, fonts once the terminal has a font pipeline.
type resourceSpec struct {
	Images []string `json:"images"`
}

// keySpec is one session-registered shortcut.
type keySpec struct {
	ID  int    `json:"id"`
	Key string `json:"key"`
}

// MenuSpec is one top-level menu from the server. The shell realizes it
// natively (NSMenu on macOS).
type MenuSpec struct {
	Title string     `json:"title"`
	Items []MenuItem `json:"items"`
}

type MenuItem struct {
	ID    int        `json:"id"`
	Title string     `json:"title"`
	Key   string     `json:"key"`
	Sep   bool       `json:"sep"`
	Items []MenuItem `json:"items"`
}

// patchOp is the union of every op shape. Absent fields stay zero.
type patchOp struct {
	Op     string            `json:"op"`
	Parent int               `json:"parent"`
	Index  int               `json:"index"`
	Node   *nodeJSON         `json:"node"`
	ID     int               `json:"id"`
	P      props             `json:"p"`
	Start  int               `json:"start"`
	Reset  bool              `json:"reset"`
	Rows   [][]string        `json:"rows"`
	Tokens map[string]string `json:"tokens"`
	// Metrics is the metric-theme op's token map ("metrics" rather than
	// "tokens": theme already owns that key in this union, with string values).
	Metrics map[string]float64 `json:"metrics"`
	Menu    []MenuSpec         `json:"menu"`
	Title   string             `json:"title"`
	Keys    []keySpec          `json:"keys"`
	Images  []string           `json:"images"`
	Meta    []metaSpec         `json:"meta"`
}

// metaSpec is one tree row's wire meta, parallel to its cells in a rows op.
type metaSpec struct {
	Key string `json:"key"`
	D   int    `json:"d"`
	K   bool   `json:"k"`
	X   bool   `json:"x"`
}

func rowMeta(in []metaSpec) []ui.RowMeta {
	if in == nil {
		return nil
	}
	out := make([]ui.RowMeta, len(in))
	for i, m := range in {
		out[i] = ui.RowMeta{Key: m.Key, Depth: m.D, Expandable: m.K, Expanded: m.X}
	}
	return out
}

type clientEvent struct {
	T     string `json:"t"`
	ID    int    `json:"id"`
	Ev    string `json:"ev"`
	Value any    `json:"value,omitempty"`
	Seq   int    `json:"seq"`
}

// Config wires a Session to its transport and shell.
type Config struct {
	URL string
	// Dialer overrides websocket.DefaultDialer, for unix sockets and in-process
	// pipes route here via NetDial.
	Dialer *websocket.Dialer
	// Wake pokes the event loop (glfw.PostEmptyEvent).
	Wake func()
	// ViewSize reports the current logical window size (for the connect URL).
	ViewSize func() (int, int)
	// SetMenu realizes a server menu spec natively. Called on the Pump
	// thread (the main thread, since Cocoa requires it). nil ignores menus.
	SetMenu func(menus []MenuSpec)
	// SetTitle renames the window (server-driven `title`). Called on the
	// Pump thread. nil ignores titles.
	SetTitle func(title string)
	// Preload warms a resource cache entry (server-driven `resource` op -
	// images today). Called on the Pump thread. nil ignores preloads.
	Preload func(src string)
}

// Session is the native terminal's protocol client: connects, applies
// mount/patch to the widget tree, ships subscribed events back, reconnects
// with backoff, resumes within the process lifetime via the sid the server
// hands out (the browser keeps it in sessionStorage, here it lives in
// memory).
//
// Network I/O runs on goroutines, and everything that touches widgets is queued
// as a closure and drained by Pump() on the main (GL) thread. Wake pokes the
// event loop so a blocked WaitEvents notices.
type Session struct {
	ui     *ui.Ui
	url    string
	dialer *websocket.Dialer
	Wake   func()
	// ViewSize reports the current logical window size (for the connect URL).
	ViewSize func() (int, int)
	setMenu  func(menus []MenuSpec)
	setTitle func(title string)
	preload  func(src string)

	queue chan func()

	mu      sync.Mutex
	conn    *websocket.Conn
	lastSeq int
	sid     string

	mounted bool
	// everMounted flips once any mount happens (main thread only). From then
	// on the tree outlives the socket, disconnects show a banner instead of
	// the status screen.
	everMounted bool
	retry       time.Duration

	resizeAt  time.Time
	resizeW   int
	resizeH   int
	hasResize bool

	store *NodeStore
}

func NewSession(u *ui.Ui, cfg Config) *Session {
	dialer := cfg.Dialer
	if dialer == nil {
		dialer = websocket.DefaultDialer
	}
	s := &Session{
		ui:       u,
		url:      cfg.URL,
		dialer:   dialer,
		Wake:     cfg.Wake,
		ViewSize: cfg.ViewSize,
		setMenu:  cfg.SetMenu,
		setTitle: cfg.SetTitle,
		preload:  cfg.Preload,
		queue:    make(chan func(), 256),
		retry:    500 * time.Millisecond,
	}
	s.store = NewNodeStore(s)
	u.OnAux = func(button int) { s.Event(0, "aux", button) }
	s.showStatus(fmt.Sprintf("connecting to %s …", cfg.URL))
	go s.connect()
	return s
}

// Pump runs queued protocol work on the caller's thread (the GL thread).
func (s *Session) Pump() {
	for {
		select {
		case fn := <-s.queue:
			fn()
		default:
			s.flushResize()
			return
		}
	}
}

func (s *Session) post(fn func()) {
	s.queue <- fn
	if s.Wake != nil {
		s.Wake()
	}
}

// -- connection ----------------------------------------------------------------------

func (s *Session) connect() {
	u, err := url.Parse(s.url)
	if err != nil {
		s.post(func() { s.showStatus("bad url: " + s.url) })
		return
	}
	q := u.Query()
	if s.ViewSize != nil {
		w, h := s.ViewSize()
		q.Set("vw", fmt.Sprint(w))
		q.Set("vh", fmt.Sprint(h))
	}
	s.mu.Lock()
	if s.sid != "" {
		q.Set("resume", s.sid)
		// Present the last seq we applied: if nothing changed server-side,
		// the resume comes back as a bare ack and our world stays put.
		q.Set("seq", fmt.Sprint(s.lastSeq))
	}
	s.mu.Unlock()
	u.RawQuery = q.Encode()

	conn, _, err := s.dialer.Dial(u.String(), nil)
	if err != nil {
		s.scheduleRetry()
		return
	}
	s.mu.Lock()
	s.conn = conn
	s.retry = 500 * time.Millisecond
	s.mu.Unlock()

	go s.readLoop(conn)
}

func (s *Session) scheduleRetry() {
	s.mu.Lock()
	d := s.retry
	s.retry = min(s.retry*8/5, 8*time.Second)
	s.mu.Unlock()
	s.post(func() {
		s.mounted = false
		// A tree we've shown stays up, visible and locally interactive
		// (scroll, selection), under a reconnect banner. Only a session
		// that never mounted falls back to the full status screen.
		if s.everMounted {
			s.showBanner("connection lost - reconnecting…")
		} else {
			s.showStatus(fmt.Sprintf("server unreachable at %s - retrying", s.url))
		}
	})
	time.AfterFunc(d, s.connect)
}

func (s *Session) readLoop(conn *websocket.Conn) {
	for {
		_, data, err := conn.ReadMessage()
		if err != nil {
			s.mu.Lock()
			current := s.conn == conn
			if current {
				s.conn = nil
			}
			s.mu.Unlock()
			if current {
				s.scheduleRetry()
			}
			return
		}
		var msg serverMsg
		if err := json.Unmarshal(data, &msg); err != nil {
			log.Println("caution: bad server message:", err)
			continue
		}
		s.post(func() { s.handle(&msg) })
	}
}

// -- message handling (main thread) ----------------------------------------------------

func (s *Session) handle(msg *serverMsg) {
	s.mu.Lock()
	s.lastSeq = msg.Seq
	s.mu.Unlock()
	switch msg.T {
	case "mount":
		if msg.SID != "" {
			s.mu.Lock()
			s.sid = msg.SID
			s.mu.Unlock()
		}
		if msg.Theme != nil {
			ui.ApplyTheme(msg.Theme)
		}
		if msg.Metrics != nil {
			s.ui.ApplyMetrics(msg.Metrics)
		}
		if msg.Menu != nil && s.setMenu != nil {
			s.setMenu(msg.Menu)
		}
		if msg.Title != "" && s.setTitle != nil {
			s.setTitle(msg.Title)
		}
		if msg.Keys != nil {
			s.applyKeys(msg.Keys)
		}
		if msg.Resources != nil {
			s.applyResources(msg.Resources.Images)
		}
		clear(s.store.ByID)
		if msg.Root != nil {
			root := s.store.Build(*msg.Root)
			s.mounted = true
			s.everMounted = true
			s.ui.SetRoot(root)
			s.hideBanner()
		}
	case "resume":
		// Nothing changed while we were away: the tree we're showing is
		// still right, keep everything.
		s.mounted = true
		s.everMounted = true
		s.hideBanner()
	case "patch":
		if !s.mounted {
			return
		}
		// Ops that name the node they touch mark just that node (the paint
		// layer splices everything around it), while ops that reshape the tree or
		// the client's chrome fall back to rebuilding the whole frame.
		full := false
		for i := range msg.Ops {
			if s.applyOp(&msg.Ops[i]) {
				full = true
			}
		}
		if full {
			s.ui.Invalidate()
		}
	}
}

// applyOp applies one op and reports whether the frame after it has to be
// rebuilt whole.
func (s *Session) applyOp(op *patchOp) bool {
	switch op.Op {
	case "set":
		if w, ok := s.store.ByID[op.ID]; ok {
			s.store.Apply(w, op.ID, op.P)
			// The protocol layer knows which node a set op touched,
			// so hand the paint layer a targeted invalidation.
			s.ui.Damage(w)
		}
	case "insert":
		parent, ok := s.store.ByID[op.Parent]
		if !ok || op.Node == nil {
			return true
		}
		s.ui.NoteStructural()
		w := s.store.Build(*op.Node)
		w.Base().Parent = parent
		pk := parent.Base()
		idx := min(op.Index, len(pk.Kids))
		pk.Kids = append(pk.Kids[:idx], append([]ui.Widget{w}, pk.Kids[idx:]...)...)
		return true
	case "remove":
		s.ui.NoteStructural()
		s.store.Remove(op.ID)
		return true
	case "rows":
		if w, ok := s.store.ByID[op.ID].(*ui.TableView); ok {
			w.ApplyRows(op.Start, op.Rows, rowMeta(op.Meta), op.Reset) // marks itself
		}
	case "theme":
		// a patched theme is a switch the user watched happen, so fade it. the
		// mount's theme still snaps, since first paint has no before
		s.ui.AnimateTheme(op.Tokens) // every widget's colors: whole frame
	case "metrics":
		// Geometry snaps (no tween) and invalidates measurement caches along
		// with the frame. See Ui.ApplyMetrics.
		s.ui.ApplyMetrics(op.Metrics)
	case "menu":
		if s.setMenu != nil {
			s.setMenu(op.Menu)
		}
		return true
	case "title":
		if s.setTitle != nil {
			s.setTitle(op.Title)
		}
	case "keys":
		s.applyKeys(op.Keys)
	case "resource":
		s.applyResources(op.Images)
	case "move":
		w, okW := s.store.ByID[op.ID]
		parent, okP := s.store.ByID[op.Parent]
		if !okW || !okP {
			return true
		}
		s.ui.NoteStructural()
		if old := w.Base().Parent; old != nil {
			ok := old.Base()
			for i, c := range ok.Kids {
				if c == w {
					ok.Kids = append(ok.Kids[:i], ok.Kids[i+1:]...)
					break
				}
			}
		}
		w.Base().Parent = parent
		pk := parent.Base()
		idx := min(op.Index, len(pk.Kids))
		pk.Kids = append(pk.Kids[:idx], append([]ui.Widget{w}, pk.Kids[idx:]...)...)
		return true
	}
	return false
}

// applyKeys installs the session's registered shortcuts on the Ui (main
// thread, since handle() runs on Pump).
func (s *Session) applyKeys(keys []keySpec) {
	m := make(map[string]int, len(keys))
	for _, k := range keys {
		m[k.Key] = k.ID
	}
	s.ui.KeyCombos = m
	s.ui.OnKeyCombo = func(id int) { s.Event(0, "key", id) }
}

func (s *Session) applyResources(images []string) {
	if s.preload == nil {
		return
	}
	for _, src := range images {
		s.preload(src)
	}
}

// -- upstream events -------------------------------------------------------------------

// Event implements EventSink. Safe from any goroutine (input debounce timers
// fire off-thread).
func (s *Session) Event(id int, ev string, value any) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.conn == nil {
		return
	}
	msg := clientEvent{T: "ev", ID: id, Ev: ev, Value: value, Seq: s.lastSeq}
	if err := s.conn.WriteJSON(msg); err != nil {
		log.Println("caution: event send failed:", err)
	}
}

// NoteResize debounces window resizes into the session-level resize event
// (node id 0, because no widget owns the viewport).
func (s *Session) NoteResize(w, h int) {
	s.resizeW, s.resizeH = w, h
	s.resizeAt = time.Now().Add(250 * time.Millisecond)
	s.hasResize = true
}

func (s *Session) flushResize() {
	if s.hasResize && time.Now().After(s.resizeAt) {
		s.hasResize = false
		s.Event(0, "resize", map[string]any{"w": s.resizeW, "h": s.resizeH})
	}
}

// -- status screen ---------------------------------------------------------------------

// showBanner puts a small client-owned pill over the live tree while
// reconnecting. The tree beneath keeps working locally, and the banner never
// intercepts input.
func (s *Session) showBanner(text string) {
	f := gfx.NewFont(12, gfx.FontOpts{Weight: 500})
	m := s.ui.Measure(f, text)
	const padL, dot, gap, padR = 12, 8, 8, 14
	root := ui.NewPanel() // transparent, full-viewport carrier
	pill := ui.NewPanel()
	root.Add(pill)
	pill.Width = f32p(float32(math.Ceil(float64(padL + dot + gap + m.Width + padR))))
	pill.Height = f32p(30)
	pill.Anchors = &ui.Anchors{CenterX: f32p(0), Top: f32p(10)}
	pill.Bg = ui.Tok("panelAlt")
	pill.Radius = gfx.CornerRadius(15)
	pill.BorderColor = ui.Tok("edge")
	pill.DropShadow = &ui.Shadow{Blur: 14, Color: gfx.WithAlpha(gfx.Black, 0.4), Dy: 3}
	dotW := ui.NewPanel()
	pill.Add(dotW)
	dotW.Width = f32p(dot)
	dotW.Height = f32p(dot)
	dotW.Radius = gfx.CornerRadius(dot / 2)
	dotW.Bg = ui.Tok("accent")
	dotW.Anchors = &ui.Anchors{Left: f32p(padL), CenterY: f32p(0)}
	label := ui.NewLabel(text, f, ui.Tok("inkDim"))
	pill.Add(label)
	label.Anchors = &ui.Anchors{Left: f32p(padL + dot + gap), CenterY: f32p(0)}
	s.ui.Banner = root
	s.ui.Invalidate()
}

func (s *Session) hideBanner() {
	if s.ui.Banner != nil {
		s.ui.Banner = nil
		s.ui.Invalidate()
	}
}

// showStatus is the minimal client-owned screen for the disconnected state.
func (s *Session) showStatus(text string) {
	root := ui.NewPanel()
	card := ui.NewPanel()
	root.Add(card)
	card.Anchors = &ui.Anchors{CenterX: f32p(0), CenterY: f32p(-40)}
	card.Width = f32p(560)
	card.Height = f32p(120)
	card.Bg = ui.Tok("panelAlt")
	card.BorderColor = ui.Tok("edgeSoft")
	card.Radius = gfx.CornerRadius(12)
	title := ui.NewLabel("caution", gfx.NewFont(18, gfx.FontOpts{Weight: 600}), nil)
	title.Anchors = &ui.Anchors{Left: f32p(24), Top: f32p(24)}
	card.Add(title)
	detail := ui.NewLabel(text, gfx.NewFont(13, gfx.FontOpts{}), ui.Tok("inkDim"))
	detail.Anchors = &ui.Anchors{Left: f32p(24), Top: f32p(58)}
	detail.Selectable = true
	card.Add(detail)
	s.ui.SetRoot(root)
}

func f32p(v float32) *float32 { return &v }

File: src/proto/session.ts
import { BLACK, hex, withAlpha } from '../gfx/color';
import { font } from '../gfx/font';
import { TableView } from '../ui/tableview';
import { bumpThemeEpoch, theme } from '../ui/theme';
import { Ui } from '../ui/ui';
import { Label, Panel } from '../ui/widgets';
import { Widget } from '../ui/widget';
import { NodeStore } from './inflate';
import type { ClientEvent, PatchOp, ServerMsg } from './types';

/**
 * Client half of the wire protocol: connects, applies mount/patch messages to
 * the widget tree, and ships subscribed semantic events back. Reconnects with
 * backoff. The server re-sends a full mount on a fresh connection.
 */
export class Session {
  private ws: WebSocket | null = null;
  private store: NodeStore;
  private lastSeq = 0;
  private retryMs = 500;
  private mounted = false;
  /** True once any mount landed - from then on the tree outlives the socket. */
  private everMounted = false;

  constructor(
    private ui: Ui,
    private url: string,
  ) {
    this.store = new NodeStore({ event: (id, ev, value) => this.sendEvent(id, ev, value) });
    ui.onAux = (button) => this.sendEvent(0, 'aux', button);
    this.showStatus(`connecting to ${url} …`);
    this.connect();
    // Viewport reporting: the connect URL carries the initial size (so it is
    // known at mount); later changes flow as a debounced session-level event
    // (node id 0 - no widget owns the viewport).
    let timer: ReturnType<typeof setTimeout> | null = null;
    window.addEventListener('resize', () => {
      if (timer) clearTimeout(timer);
      timer = setTimeout(() => {
        timer = null;
        this.sendEvent(0, 'resize', { w: window.innerWidth, h: window.innerHeight });
      }, 250);
    });
  }

  private connect(): void {
    // Session resume: the server keeps a disconnected session's state briefly;
    // presenting its id on reconnect (or after a page reload, sessionStorage
    // survives F5, not new tabs) reattaches instead of starting over.
    const sid = sessionStorage.getItem('caution:sid');
    const q = new URLSearchParams({
      vw: String(window.innerWidth),
      vh: String(window.innerHeight),
    });
    if (sid) {
      q.set('resume', sid);
      // Present the last seq we applied: if nothing changed server-side, the
      // resume comes back as a bare ack and our whole world stays put.
      q.set('seq', String(this.lastSeq));
    }
    const ws = new WebSocket(`${this.url}?${q}`);
    this.ws = ws;
    ws.onopen = () => {
      this.retryMs = 500;
    };
    ws.onmessage = (e) => {
      const msg = JSON.parse(e.data as string) as ServerMsg;
      this.handle(msg);
    };
    ws.onclose = () => {
      if (this.ws !== ws) return;
      this.ws = null;
      this.mounted = false;
      // A tree we've shown stays up, visible and locally interactive
      // (scroll, selection), under a reconnect banner. Only a session that
      // never mounted falls back to the full status screen.
      if (this.everMounted) this.showBanner('connection lost - reconnecting…');
      else this.showStatus(`server unreachable at ${this.url} - retrying`);
      setTimeout(() => this.connect(), this.retryMs);
      this.retryMs = Math.min(this.retryMs * 1.6, 8000);
    };
  }

  private handle(msg: ServerMsg): void {
    this.lastSeq = msg.seq;
    if (msg.t === 'mount') {
      if (msg.sid) sessionStorage.setItem('caution:sid', msg.sid);
      if (msg.theme) this.applyTheme(msg.theme);
      if (msg.metrics) this.ui.applyMetrics(msg.metrics);
      if (msg.title) document.title = msg.title;
      if (msg.keys) this.applyKeys(msg.keys);
      this.applyMenu(msg.menu ?? []);
      if (msg.resources?.images) for (const src of msg.resources.images) this.ui.preloadImage(src);
      this.store.byId.clear();
      const root = this.store.build(msg.root);
      this.mounted = true;
      this.everMounted = true;
      this.ui.setRoot(root);
      this.hideBanner();
      // headless capture (go/cmd/goldens) waits for this before screenshotting
      ((window as any).__caution ??= { frames: 0, mounted: false }).mounted = true;
    } else if (msg.t === 'resume') {
      // nothing changed while we were away: the tree we're showing is right
      this.mounted = true;
      this.everMounted = true;
      this.hideBanner();
    } else if (msg.t === 'patch' && this.mounted) {
      // ops that name the node they touch mark just that node (the paint layer
      // splices everything around it). ops that reshape the tree or the
      // client's chrome fall back to rebuilding the whole frame.
      let full = false;
      for (const op of msg.ops) if (this.applyOp(op)) full = true;
      if (full) this.ui.invalidate();
    }
  }

  /** Apply one op. True if the frame after it has to be rebuilt whole. */
  private applyOp(op: PatchOp): boolean {
    if (op.op === 'set') {
      const w = this.store.byId.get(op.id);
      if (w) {
        this.store.apply(w, op.id, op.p);
        this.ui.damage(w);
      }
    } else if (op.op === 'insert') {
      const parent = this.store.byId.get(op.parent);
      if (!parent) return true;
      this.ui.noteStructural();
      const w = this.store.build(op.node);
      w.parent = parent;
      parent.children.splice(Math.min(op.index, parent.children.length), 0, w);
      return true;
    } else if (op.op === 'remove') {
      this.ui.noteStructural();
      this.store.remove(op.id);
      return true;
    } else if (op.op === 'rows') {
      const w = this.store.byId.get(op.id);
      if (w instanceof TableView) w.applyRows(op.start, op.rows, op.meta, op.reset ?? false); // marks itself
    } else if (op.op === 'theme') {
      this.applyTheme(op.tokens); // every widget's colors: whole frame
    } else if (op.op === 'metrics') {
      // geometry snaps (no tween) and invalidates measurement caches along
      // with the frame
      this.ui.applyMetrics(op.metrics);
    } else if (op.op === 'title') {
      document.title = op.title;
    } else if (op.op === 'keys') {
      this.applyKeys(op.keys);
    } else if (op.op === 'menu') {
      this.applyMenu(op.menu ?? []);
      return true;
    } else if (op.op === 'resource') {
      for (const src of op.images ?? []) this.ui.preloadImage(src);
    } else if (op.op === 'move') {
      const w = this.store.byId.get(op.id);
      const parent = this.store.byId.get(op.parent);
      if (!w || !parent) return true;
      this.ui.noteStructural();
      if (w.parent) {
        const i = w.parent.children.indexOf(w);
        if (i >= 0) w.parent.children.splice(i, 1);
      }
      w.parent = parent;
      parent.children.splice(Math.min(op.index, parent.children.length), 0, w);
      return true;
    }
    return false;
  }

  private applyKeys(keys: { id: number; key: string }[]): void {
    this.ui.setKeyCombos(
      new Map(keys.map((k) => [k.key, k.id])),
      (id) => this.sendEvent(0, 'key', id),
    );
  }

  /** Realize the server's menu spec as the in-window menu bar. Picks are
   * session-level `menu` events like NSMenu picks natively. */
  private applyMenu(menu: { title: string; items: import('./types').MenuItemSpec[] }[]): void {
    this.ui.setMenubar(menu, (id) => this.sendEvent(0, 'menu', id));
  }

  private themeTween: number | null = null;

  /**
   * Retarget theme tokens. Widgets hold references to token Colors resolved
   * at inflate time, so tokens always mutate in place. The first application
   * (mount) snaps, since first paint has no before, and later ones fade over
   * ~180ms.
   */
  private applyTheme(tokens: Record<string, string>): void {
    const targets: { c: { r: number; g: number; b: number; a: number }; from: number[]; to: number[] }[] = [];
    for (const [k, v] of Object.entries(tokens)) {
      const c = hex(v);
      const existing = theme[k];
      if (existing) {
        targets.push({ c: existing, from: [existing.r, existing.g, existing.b, existing.a], to: [c.r, c.g, c.b, c.a] });
      } else {
        theme[k] = c; // brand-new tokens have nothing to fade from
      }
    }
    const snap = (): void => {
      for (const { c, to } of targets) {
        [c.r, c.g, c.b, c.a] = to as [number, number, number, number];
      }
      bumpThemeEpoch();
    };
    if (this.themeTween != null) {
      cancelAnimationFrame(this.themeTween);
      this.themeTween = null;
    }
    if (!this.everMounted || !targets.length) {
      snap();
      return;
    }
    const start = performance.now();
    const dur = 180;
    const step = () => {
      const t = Math.min(1, (performance.now() - start) / dur);
      const e = t * (2 - t); // ease-out
      for (const { c, from, to } of targets) {
        c.r = from[0]! + (to[0]! - from[0]!) * e;
        c.g = from[1]! + (to[1]! - from[1]!) * e;
        c.b = from[2]! + (to[2]! - from[2]!) * e;
        c.a = from[3]! + (to[3]! - from[3]!) * e;
      }
      bumpThemeEpoch();
      this.ui.invalidate();
      this.themeTween = t < 1 ? requestAnimationFrame(step) : null;
    };
    step();
  }

  private sendEvent(id: number, ev: string, value?: unknown): void {
    if (this.ws?.readyState !== WebSocket.OPEN) return;
    const msg: ClientEvent = { t: 'ev', id, ev, seq: this.lastSeq };
    if (value !== undefined) msg.value = value;
    this.ws.send(JSON.stringify(msg));
  }

  /**
   * Small client-owned pill over the live tree while reconnecting. The tree
   * beneath keeps working locally.
   */
  private showBanner(text: string): void {
    const f = font(12, { weight: 500 });
    const m = this.ui.measure(f, text);
    const padL = 12;
    const dot = 8;
    const gap = 8;
    const padR = 14;
    const root = new Panel(); // transparent, full-viewport carrier
    const pill = root.add(new Panel());
    pill.width = padL + dot + gap + Math.ceil(m.width) + padR;
    pill.height = 30;
    pill.anchors = { centerX: 0, top: 10 };
    pill.bg = theme.panelAlt;
    pill.radius = 15;
    pill.borderColor = theme.edge;
    pill.dropShadow = { blur: 14, color: withAlpha(BLACK, 0.4), offset: [0, 3] };
    const dotW = pill.add(new Panel());
    dotW.width = dot;
    dotW.height = dot;
    dotW.radius = dot / 2;
    dotW.bg = theme.accent;
    dotW.anchors = { left: padL, centerY: 0 };
    const label = pill.add(new Label(text, f, theme.inkDim));
    label.anchors = { left: padL + dot + gap, centerY: 0 };
    this.ui.banner = root;
    this.ui.invalidate();
  }

  private hideBanner(): void {
    if (this.ui.banner) {
      this.ui.banner = null;
      this.ui.invalidate();
    }
  }

  /** Minimal client-owned screen for the disconnected state. */
  private showStatus(text: string): void {
    const root: Widget = new Panel();
    const card = root.add(new Panel());
    card.anchors = { centerX: 0, centerY: -40 };
    card.width = 560;
    card.height = 120;
    card.bg = theme.panelAlt;
    card.radius = 12;
    card.borderColor = theme.edgeSoft;
    const title = card.add(new Label('caution', font(18, { weight: 600 })));
    title.anchors = { left: 24, top: 24 };
    const detail = card.add(new Label(text, font(13), theme.inkDim));
    detail.anchors = { left: 24, top: 58 };
    detail.selectable = true;
    this.ui.setRoot(root);
  }
}
